src/utils/link-channel.js

```diff
@@ -0,0 +1,141 @@
+import { generatePublicKeyAndEncryptedObject, generateKeys, decryptObject } from "./crypto";
+
+export default class LinkChannel {
+  constructor(store) {
+    this.store = store;
+  }
+
+  setSocket = socket => {
+    this.socket = socket;
+  };
+
+  // Returns a promise that, when resolved, will forward an object with three keys:
+  //
+  // code: The code that was made available to use for link.
+  //
+  // cancel: A function that the caller can call to cancel the use of the code.
+  //
+  // onFinished: A promise that, when resolved, indicates the code is no longer usable,
+  // because it was either successfully used by the remote device or it has expired
+  // ("used" or "expired" is passed to the callback.)
+  generateCode = () => {
+    return new Promise(resolve => {
+      const onFinished = new Promise(finished => {
+        const step = () => {
+          const code = Math.floor(Math.random() * 9999)
+            .toString()
+            .padStart(4, "0");
+
+          // Only respond to one link_request in this channel.
+          let readyToSend = false;
+          let leftChannel = false;
+
+          const channel = this.socket.channel(`link:${code}`, { timeout: 10000 });
+
+          const leave = () => {
+            if (!leftChannel) channel.leave();
+            leftChannel = true;
+          };
+
+          const cancel = () => leave();
+
+          channel.on("link_expired", () => finished("expired"));
+
+          channel.on("presence_state", state => {
+            if (readyToSend) return;
+
+            if (Object.keys(state).length > 0) {
+              // Code is in use by someone else, try a new one
+              step();
+            } else {
+              readyToSend = true;
+              resolve({ code, cancel, onFinished });
+            }
+          });
+
+          channel.on("link_request", incoming => {
+            if (readyToSend) {
+              const data = { path: location.pathname };
+
+              // Copy profile data to link'ed device if it's been set.
+              if (this.store.state.activity.hasChangedName) {
+                data.profile = { ...this.store.state.profile };
+              }
+
+              generatePublicKeyAndEncryptedObject(incoming.public_key, data).then(
+                ({ publicKeyString, encryptedData }) => {
+                  const payload = {
+                    target_session_id: incoming.reply_to_session_id,
+                    public_key: publicKeyString,
+                    data: encryptedData
+                  };
+
+                  if (!leftChannel) {
+                    channel.push("link_response", payload);
+                  }
+
+                  leave();
+
+                  finished("used");
+                  readyToSend = false;
+                }
+              );
+            }
+          });
+
+          channel.join().receive("error", r => console.error(r));
```

Comment:
You actually don't want to reject here, because I'm pretty sure this can happen for any error in the underlying transport layer, including ones that are recoverable. Specifically, if we deploy reticiulum when someone is in this routine, their underlying websocket will be disconnected and reconnected, but an error will be thrown on the channel. (But, it should gracefully recover I believe)

Reply:
oh good call. do you think it's perhaps worth capturing these errors in Sentry just in case to better understand why, when, and how often this happens.